### PR TITLE
Clean up package structure

### PR DESCRIPTION
**Pull request by joschi**
In order to properly support the Java Platform Module System (JPMS), the packages in modules must not overlap, or put differently, the packages may not be split into multiple modules.

See also: https://nipafx.dev/java-9-migration-guide/#split-packages

This change set cleans up the overlapping package structure where necessary (and is unfortunately breaking backward compatibility).

Affected packages:

- `dropwizard-core`: `io.dropwizard` -> `io.drowizard.core`
- `dropwizard-views`: `io.dropwizard.views` -> `io.dropwizard.views.common`
- `dropwizard-metrics`: `io.dropwizard.metrics` -> `io.dropwizard.metrics.common`
- `dropwizard-logging`: `io.dropwizard.logging` -> `io.dropwizard.logging.common`

Refs https://github.com/dropwizard/dropwizard/pull/2926
Refs https://github.com/dropwizard/dropwizard/issues/3131
Refs https://github.com/dropwizard/dropwizard/issues/3654

### CODE REVIEW

**Review by nickbabcock — DISMISSED** [review]
Yeah looks like the structure needed to be cleaned up to be compliant. Thanks for this and making it backwards compatible.


You also mentioned


> We'll need more tests to see, if everything works with Java 9+ ; when using the module path


I'm not sure how one would test the module path. Were you able to include them in the PR?


> `io.dropwizard.metrics` -> `io.dropwizard.metrics.{common, core}`


I'd say, unless you want to make everything use the `core` suffix, having `io.dropwizard.metrics` match logging and views with the `common` suffix make sense

**Comment by pkwarren** [comment]
I'm worried that the actual changes aren't going to be backwards compatible - if we end up changing the locations of classes it could break consumers.

For example, by changing contract to contract superclass (i.e. from io.dropwizard.Configuration to io.dropwizard.core.Configuration), we may break compatibility. http://wiki.eclipse.org/Evolving_Java-based_APIs_2 is a great reference for these sorts of changes to know the impact. We can also run a tool like https://lvc.github.io/japi-compliance-checker/ to verify the API contract hasn't changed. OkHttp also just updated from Java to Kotlin (https://square.github.io/okhttp/upgrading_to_okhttp_4/) and recommended use of https://github.com/siom79/japicmp to verify no API changes.

If we can't make this backwards compatible, then we might want to consider this for a 3.x release and change groupId, package name, and introduce compat packages if we want to support existing codebases. For example:

* Add io.dropwizard:dropwizard3-core which includes package io.dropwizard3.*
* Add io.dropwizard:dropwizard3-core-compat artifact which includes all classes from dropwizard-core 2.x package which is deprecated and extend dropwizard3-core classes.
* Update io.dropwizard:dropwizard-core to pull in dropwizard3-core and dropwizard3-core-compat artifacts. This way existing dependencies work unchanged.

**Comment by joschi** [comment]
After playing around with this branch a little bit more I couldn't come up with a way to _not_ break backward compatibility in one way or the other.

So I think it makes sense to rip off the bandaid and introduce the breaking change in Dropwizard 2.1.0 to be done with it.

Going forward, this would mean that Dropwizard could be compatible to the JPMS and no other package renaming would have to be done.

The other possibility would be to postpone this change to Dropwizard 3.0.0 and do it together with the move to the Jakarta EE namespace (Jetty 11, Jersey 3, Hibernate Validator 7, etc.) and the Java 11 baseline (or drop this altogether).

@dropwizard/committers Any opinions?

**Comment by jplock** [comment]
@joschi I think this would surprise people in 2.1.0 even if we heavily documented it.  I'd be in favor of starting 3.x with the changes you proposed and maybe at the same time saying 1.x would no longer be maintained/supported (even bug fixes), to encourage people to upgrade to at least 2.x.

**Comment by joschi** [comment]
> and maybe at the same time saying 1.x would no longer be maintained/supported

@jplock Thankfully we already did that with [Dropwizard 1.3.29](https://github.com/dropwizard/dropwizard/releases/tag/v1.3.29). 😃

I don't have a strong opinion on whether to to this change in Dropwizard 2.1.x or 3.0.x.

The only thing which might not be desirable is mixing too many breaking changes into Dropwizard 3.0.x.

**Comment by pstackle** [comment]
I'll put in a vote for holding off the change until a 3.0 release. Having breaking changes is much more expected for a major version bump compared to a minor version bump.

**Comment by zUniQueX** [comment]
> The only thing which might not be desirable is mixing too many breaking changes into Dropwizard 3.0.x.

Thanks to our current upgrade plans, Dropwizard 3.0.x won't introduce that much API changes. Since Jersey will bring JPMS support firstly in 3.1.0 on the `jakarta` namespace, we don't even have the need to bring this to DW 3. But I see no problem bringing this to the 3.0.0 release.

**Comment by joschi** [comment]
@dropwizard/committers I've rebased the branch on `release/3.0.x`.

File: docs/source/getting-started.rst
Comment (zUniQueX) [inline docs/source/getting-started.rst:283]: Overall looks good, but some doc imports seem to be broken

Comment (joschi) [reply docs/source/getting-started.rst:283]: @zUniQueX Thanks for noticing! 😄

File: dropwizard-logging/src/main/java/io/dropwizard/logging/common/filter/FilterFactory.java
Comment (zUniQueX) [inline dropwizard-logging/src/main/java/io/dropwizard/logging/common/filter/FilterFactory.java:15]: @joschi Is this corrent in the new implementation?

**Review by zUniQueX — APPROVED** [review]
@joschi This looks very good, thanks for the effort :heart:

I just have a few minor suggestions:
 - `DefaultServerFactoryTest:18-19` multiple imports for `Environment` and `ExceptionMapperBinder`
 - `SimpleServerFactoryTest:17` multiple imports for `Environment`
 - `JdbiTest:8` multiple imports for `Environment`
 - `getting-started.rst:196` broken import
 - `configuration.rst:892` package `logging.common` instead of `logging.filter.common`
 - `FilterFactory:15` broken JavaDoc
 
And one question left: Could it get a problem, that the packages in `dropwizard-e2e` and `dropwizard-example` are named `com.example` and `com.example.helloworld`?

**Comment by joschi** [comment]
@zUniQueX Thanks for the detailed look at this PR! ❤️ 

> And one question left: Could it become a problem, that the packages in `dropwizard-e2e` and `dropwizard-example` are named `com.example` and `com.example.helloworld`?

No, these are not an issue because neither is being published and consumed by other projects and they never end up in the same application and module path.

**Comment by zUniQueX** [comment]
> > And one question left: Could it become a problem, that the packages in `dropwizard-e2e` and `dropwizard-example` are named `com.example` and `com.example.helloworld`?
> 
> No, these are not an issue because neither is being published and consumed by other projects and they never end up in the same application and module path.

Okay, thanks for taking a look at this :+1:

**Pull request merged**
